packages/request/lib/SessionPool.js

```diff
@@ -0,0 +1,274 @@
+
+const delay = milliseconds => new Promise(resolve => setTimeout(resolve, milliseconds))
+
+class SessionPool {
+  constructor (id) {
+    this.id = id
+    this.sessions = new Set()
+    this.tlsSession = undefined
+    this.queue = []
+  }
+
+  get options () {
+    return this.id.getOptions()
+  }
+
+  get pingInterval () {
+    return this.options.pingInterval
+  }
+
+  get keepAliveTimeout () {
+    return this.options.keepAliveTimeout
+  }
+
+  get connectTimeout () {
+    return this.options.connectTimeout
+  }
+
+  get value () {
+    let value = 0
+    for (const { [kSemaphore]: semaphore } of this.sessions) {
+      value += semaphore.value
+    }
+    return value
+  }
+
+  destroy () {
+    this.tlsSession = undefined
+    for (const session of this.sessions) {
+      this.deleteSession(session)
+    }
+  }
+
+  addTask (resolve, reject) {
+    let settled = false
+    const callback = (err, session) => {
+      if (!settled) {
+        settled = true
+        clearTimeout(timeoutId)
+        if (err) {
+          reject(err)
+        } else {
+          resolve(session)
+        }
+      }
+    }
+    const timeoutId = setTimeout(() => {
+      const index = this.queue.indexOf(callback)
+      if (index > -1) {
+        this.queue.splice(index, 1)
+      }
+      callback(new TimeoutError(`No session could be aquired within ${this.connectTimeout} ms`))
+    }, this.connectTimeout)
+    this.queue.push(callback)
+  }
+
+  acquire () {
+    const ticket = new Promise((resolve, reject) => this.addTask(resolve, reject))
+    this.dispatch()
+    return ticket
+  }
+
+  async dispatch () {
+    let scalingUp = false
+    const scaleUp = async () => {
+      if (!scalingUp) {
+        scalingUp = true
+        try {
+          await this.createSession()
+        } catch (err) {
+          const callback = this.queue.shift()
+          callback(err)
+          const milliseconds = Math.floor((2 * Math.random() - 1) * 25) + 100
+          await delay(milliseconds)
+        } finally {
+          scalingUp = false
+          this.dispatch()
+        }
+      }
+    }
+
+    while (this.queue.length) {
+      const session = this.find()
+      if (!session) {
+        scaleUp()
+        break
+      }
+      const callback = this.queue.shift()
+      const release = await session[kSemaphore].acquire()
+      callback(null, [session, release])
+    }
+  }
+
+  find () {
+    const sessionList = Array.from(this.sessions)
+    return sessionList
+      // consider free sessions only
+      .filter(({ [kSemaphore]: semaphore }) => {
+        return semaphore.value > 0
+      })
+      // session with the highest load first
+      .sort(({ [kSemaphore]: a }, { [kSemaphore]: b }) => {
+        return a.value - b.value
+      })
+      .shift()
+  }
+
+  async request (headers, options) {
+    const [session, release] = await this.acquire()
+    clearTimeout(session[kTimeoutId])
+    const stream = session.request(headers, options)
```

Comment:
make sure that release is called in every case even if the request failed or stream does not become ready or any other unexpected situation

Reply:
Add session pool usage info to log

Reply:
✔️